x-pack/plugins/enterprise_search/server/lib/enterprise_search_request_handler.ts

```diff
@@ -126,6 +130,36 @@ export class EnterpriseSearchRequestHandler {
     };
   }
 
+  /**
+   * This path helper is similar to React Router's generatePath, but much simpler &
+   * does not use regexes. It enables us to pass a static '/foo/:bar/baz' string to
+   * createRequest({ path }) and have :bar be automatically replaced by the value of
+   * request.params.bar.
+   * It also (very importantly) wraps all URL request params with encodeURIComponent(),
+   * which is an extra layer of encoding required by the Enterprise Search server in
+   * order to correctly & safely parse user-generated IDs with special characters in
+   * their names - just encodeURI alone won't work.
+   */
+  encodePathParams(path: string, params: UnknownObject) {
+    const hasParams = path.includes(':');
+    if (!hasParams) {
+      return path;
```

Comment:
Same as above (below?) comment, paths without dynamic vars (e.g. just a plain `/hello/world` route) aren't encoded because I figure we hopefully know what we're doing. But if people feel strongly otherwise I can wrap this in `encodeURI(path)`

Reply:
> user-generated IDs

I'm surprised we don't explicitly limit what a user can actually use as an ID to disallow special characters in Ids.

Reply:
Ya same, no idea why here either but I'm guessing there's a business reason 🤔 I do remember Orhan fixing an issue reported by a customer that involved encoding/decoding document IDs though, so there's definitely folks out there with special IDs